### PR TITLE
FEAT-#6908: Remove the warning regarding engine initialization

### LINKED ISSUE #6908: Make it clearer which parameters to pass into the engine initialization regarding number of workers
We have the following warnings when initalizing an engine.

**Ray**

```bash
UserWarning: Ray execution environment not yet initialized. Initializing...
To remove this warning, run the following python code before doing dataframe operations:
import ray

ray.init()
```

**Dask**

```bash
UserWarning: Dask execution environment not yet initialized. Initializing...
To remove this warning, run the following python code before doing dataframe operations:
from distributed import Client

client = Client()
```

By default, Ray and Dask use all available cores to create workers for parallel computation but it is not always beneficial to spawn the number of workers equal to the number of CPUs. We should make it clearer which parameters to pass into the engine initialization regarding a number of workers to spawn. It would be especially useful for Dask, which creates both worker threads and processes by default.

### PR DESCRIPTION
**Pull request by YarShev**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6908  
- [x] tests passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/execution/dask/common/utils.py [hunk 1]
```diff
@@ -59,7 +59,11 @@ def _disable_warnings():
 
         # when the client is initialized, environment variables are inherited
         with set_env(PYTHONWARNINGS="ignore::FutureWarning"):
-            client = Client(n_workers=num_cpus, memory_limit=worker_memory_limit)
+            client = Client(
+                n_workers=num_cpus,
+                threads_per_worker=1,
```
Comment (anmyachev) [inline modin/core/execution/dask/common/utils.py:64]: Why?

Comment (YarShev) [reply modin/core/execution/dask/common/utils.py:64]: On the benchmark I am measuring read_csv is significantly slower with the default client rather than with the configured one. But we can get it in a separate PR as the change is not related to this PR.

```bash
read_csv: 13.72040319442749  # Client()
read_csv: 3.934723377227783  # Client(n_workers=16, threads_per_worker=1)
```

File: modin/core/execution/dask/common/utils.py [hunk 2]
```diff
@@ -59,7 +59,11 @@ def _disable_warnings():
...
+                memory_limit=worker_memory_limit,
+            )
```
Comment (YarShev) [inline modin/core/execution/dask/common/utils.py:66]:
```suggestion
            client = Client(n_workers=num_cpus, memory_limit=worker_memory_limit)
```

File: modin/core/execution/dask/common/utils.py [hunk 3]
```diff
@@ -50,7 +50,7 @@ def _disable_warnings():
             """
     from distributed import Client
 
-    client = Client()
+    client = Client(n_workers=<N1>, threads_per_worker=<N2>)
```
Comment (anmyachev) [inline modin/core/execution/dask/common/utils.py:53]: Don't you think this is an additional complication for users?

Let's also discuss warning in general.

Don't you think that such a warning encourages users to initialize the engine themselves (a common reaction to avoid annoying warnings), which in many situations is worse than what Modin does itself.

```python
>>> import modin.pandas as pd
>>> pd.DataFrame([1,2,3,4]) 
UserWarning: Ray execution environment not yet initialized. Initializing...
To remove this warning, run the following python code before doing dataframe operations:

    import ray
    ray.init()

2024-02-04 16:18:03,686 INFO worker.py:1673 -- Started a local Ray instance.
UserWarning: Distributing <class 'list'> object. This may take some time.
   0
0  1
1  2
2  3
3  4
```

Comment (YarShev) [reply modin/core/execution/dask/common/utils.py:53]: This might look like extra burden for the user but we intentionally emit this kind of a warning to emphasize that the engine can be initialized outiside Modin and it in turn can connent to the engine. I would rephrase the warning a bit. What do you think on this?

```bash
UserWarning: Ray execution environment not yet initialized. Initializing...
To remove this warning you may want to initialize the engine on your own
with the specific parameters passed in:

import ray

ray.init(num_cpus=<N>, ...)
```

Comment (dchigarev) [reply modin/core/execution/dask/common/utils.py:53]: > the warning encourages users to initialize the engine themselves, which in many situations is worse than what Modin does itself

I also agree with this. For ray, it could be reflected with memory problems if a user would try to initialize the engine themself without any arguments:
```python
import ray
ray.init() # only allowed to use 30% of /dev/shm's memory

# -------------------------------
import modin.pandas as pd
pd.DataFrame(...) # allowed to use 60% of /dev/shm's memory
``` 
We could have added additional instructions to the warning on how to specify the proper amount of memory, but IMO that would make the warning unreadable and not understandable, as the whole process of specifying the memory is quite complicated:
```python
# in order to initialize Ray with the same amount of object's store memory
# as with Modin, there's too much to consider for a user IMO
import ray
# at first, I have to know that my system have 16gb of shared memory,
# and then I also have to do math homework
ray.init(object_store_memory=16 * 1024 * 1024 * 1024 * 0.6)
```

So, I'm siding here with @anmyachev and wondering whether we should keep the warning at all or it's better to remove it

Comment (dchigarev) [reply modin/core/execution/dask/common/utils.py:53]: In order to put my question simpler:
1. What are the use cases where a user need to materialize the engine on their own? Modin have a lot of env variables that can control the arguments of the engine (`MODIN_CPUS`, `MODIN_MEMORY`, `MODIN_CLUSTER_PASSWORD`, ...), aren't they enough?
2. If cases where it's absolutely necessary to initialize the engine on your own are rather exceptions (as you can control pretty much everything via modin config), is it worth printing a 7-line warning all the time?

Comment (YarShev) [reply modin/core/execution/dask/common/utils.py:53]: > 1. What are the use cases where a user need to initialize the engine on their own? Modin have a lot of env variables that can control the arguments of the engine (`MODIN_CPUS`, `MODIN_MEMORY`, `MODIN_CLUSTER_PASSWORD`, ...), aren't they enough?

The use of the Modin configs might be a little more cumbersome as opposed to initializing the engine by users themselves. They may want to initailize the engine on their own with the specific configurations, which may also not exist in Modin.

> If cases where it's absolutely necessary to initialize the engine on your own are rather exceptions (as you can control pretty much everything via modin config), is it worth printing a 7-line warning all the time?

```bash
UserWarning: Ray execution environment not yet initialized. Initializing...
You may want to initialize Ray on your own with the specific configuration.
Refer to Ray documentation for this purpose on more details.
```

What do you think on this?

Comment (YarShev) [reply modin/core/execution/dask/common/utils.py:53]: I want to prevent users from initializing an engine with either `ray.init()` or `Client()`. Particularly, I am worried about Dask, which is much slower as said [earlier](https://github.com/modin-project/modin/pull/6909#discussion_r1477127503) with the default client. I think it would be good to emit a warning like this.

```bash
UserWarning: Ray execution environment not yet initialized. Initializing...
You may want to initialize Ray on your own with the specific configuration.
Refer to Ray documentation for this purpose on more details.
```

If the user is fair enough aware of the engine to be initialized, he/she can look at the engine docs and choose the necessary parameters for the initialization.

Comment (anmyachev) [reply modin/core/execution/dask/common/utils.py:53]: > This looks better to me, compared with what we have now. However, I'm still skeptical regarding having this warning at all, since in 90%+ use cases you don't need to initialize on your own. Though, I'm not going to block the decision of keeping the warning, let's wait for @anmyachev opinion

Agree. 

> I want to prevent users from initializing an engine with either ray.init() or Client(). Particularly, I am worried about Dask, which is much slower as said https://github.com/modin-project/modin/pull/6909#discussion_r1477127503 with the default client. I think it would be good to emit a warning like this.

We can write about this in our documentation, in the section on advanced usage (and remove this warning). Also, instead of warning, we can do logging, for example, of the info level.

Comment (YarShev) [reply modin/core/execution/dask/common/utils.py:53]: I am fine by removing the warning. Will add a section in the docs.

Comment (YarShev) [reply modin/core/execution/dask/common/utils.py]: Done

File: docs/usage_guide/advanced_usage/modin_engines.rst [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  client = Client(n_workers=<N1>, threads_per_worker=<N2>)
+  modin_cfg.Engine.put("dask") # # Modin will use Dask engine
+  modin_cfg.CpuCount.put(<N1>)
+
+To get more details on all possible parameters for initialization refer to `Dask Distributed documentation`_.
+
+MPI through unidist
+-------------------
+
+You can initialize MPI thought unidist engine with a specific number of CPUs (worker processes) to perform computation.
```
Comment (YarShev) [inline docs/usage_guide/advanced_usage/modin_engines.rst:46]:
```suggestion
You can initialize MPI through unidist engine with a specific number of CPUs (worker processes) to perform computation.
```

File: examples/jupyter/integrations/NLTK.ipynb [hunk 5]
```diff
@@ -29,96 +29,9 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 2,
+   "execution_count": null,
    "metadata": {},
-   "outputs": [
```
Comment (anmyachev) [inline examples/jupyter/integrations/NLTK.ipynb:34]: Why did you clear the output for jupyter notebook?

Comment (YarShev) [reply examples/jupyter/integrations/NLTK.ipynb:34]: We could only remove the concrete cell output but in a single cell there is other information about initialization like.

```bash
2023-04-06 12:15:19,701	INFO worker.py:1553 -- Started a local Ray instance.

UserWarning: Distributing <class 'pandas.core.frame.DataFrame'> object. This may take some time.
```

And this also looks to me a little inconsistent when some cells are executed but other aren't. I would leave all cells not executed.

Comment (anmyachev) [reply examples/jupyter/integrations/NLTK.ipynb:34]: I think as long as Modin produces this output, we can leave it there. However, we should also consider not showing such low-level information, at least by default.

> And this also looks to me a little inconsistent when some cells are executed but other aren't. I would leave all cells not executed.

Why? Is it unusual for a cell to not produce output in some cases?

Comment (YarShev) [reply examples/jupyter/integrations/NLTK.ipynb:34]: > I think as long as Modin produces this output, we can leave it there. However, we should also consider not showing such low-level information, at least by default.

We can't remove a part of the cell output.

> Why? Is it unusual for a cell to not produce output in some cases?

I mean that some cells will be marked as [], i.e., not executed, but others will be marked as [<n>], i.e., executed, which looks inconsistent to me.

Comment (YarShev) [reply examples/jupyter/integrations/NLTK.ipynb:34]: We can do that in a separate issue when we have time. I believe we may even face other issues with interoperability of Modin and those packages. The notebooks were added a while ago and we don't test them.

Comment (anmyachev) [reply examples/jupyter/integrations/NLTK.ipynb:34]: > We can do that in a separate issue when we have time. 

Should we not change them at all until then?

Comment (YarShev) [reply examples/jupyter/integrations/NLTK.ipynb:34]: I do not see an important reason of why we should not clear the cells output. We just say about interoperabilty (if it is there or not) in the notebooks. Maybe we should even recheck every notebooks on a regular basis to make sure the things are working as expected.

**Review by anmyachev — APPROVED** [review]
LGTM, @dchigarev?

**Review by dchigarev — APPROVED** [review]
